README.md
# Rack::Joint

Rack::Joint is a [Rack](https://github.com/rack/rack) middleware which helps to redirect with Ruby DSL.

### Requirements
Rack::Joint requires Ruby 2.4+.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'rack-joint'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install rack-joint

## Usage

Write settings in `config.ru`. This is a sample:

```ruby
# config.ru
require 'rack/joint'

use Rack::Joint do
  host 'example.com' do
    # If you access `http://example.com/dogs/bark.html`,
    # redirect to `http://example.org/bowwow.html` with 301.
    redirect '/dogs/bark.html' do
      new_host 'example.org'
      to '/bowwow.html'
      status 301
    end

    # If you access `http://example.com/cats.html`,
    # redirect to `https://example.org/meow/mew` with 302.
    redirect '/cats.html' do
      new_host 'example.org'
      ssl true
      to '/meow/mew'
      status 302
    end
  end

  host 'example.net' do
    # If you access `http://example.net/frog.html`,
    # redirect to `http://example.net/croak` with 301.
    redirect '/frog.html' do
      to '/croak'
    end

    # If you access `http://example.net/quack`,
    # redirect to `http://example.org/quack` with 301.
    redirect '/quack' do
      new_host 'example.org'
    end
  end

  # Following description allows you to redirect to URL you set with only host replaced.
  #
  # e.g. When accessing 'http://example.io/foo/bar/baz',
  #      redirect to `http://example.org/foob/bar/baz` with 301.
  host 'example.io' do
    redirect do
      new_host 'example.org'
    end
  end
end

run lambda { |env| [200, {'Content-Type' => 'text/plain'}, ['Hello World!']] }
```

You can use following resources with block.

| Resource | Type | Description |
| :------- | :--- | :---------- |
| `host`   | `string` | Required. Host name. |
| `redirect` | `string` | Path name. If you give nothing with this option, Rack::Joint will replace only host name. |
| `ssl` | `boolean` | Whether to enable SSL. If the option isn't set, the scheme of `Location` header is determined in response to GET request. |
| `new_host` | `string` | A new host name redirects to. |
| `to` | `string` | A new path name redirects to. |
| `status` | `integer` | Status when redirecting. You can use `301`, `302`, `303`, `307`, `308`; which is `301` to default. |

#### Supplement

[This library's test case](https://github.com/akito19/rack-joint/tree/master/test/config) will be actual samples.

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake test` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/akito19/rack-joint.

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


CHANGELOG.md
# CHANGELOG

## master

### 0.3.5
* Add Ruby 3.1 and 3.2 test and specify Rack version. [#21](https://github.com/akito19/rack-joint/pull/21)

### 0.3.4 (2019-03-25)
* Allow query parameters when request URLs have them. [#18](https://github.com/akito19/rack-joint/pull/18)

### 0.3.3 (2019-01-19)
* Support Bundler2. [#15](https://github.com/akito19/rack-joint/pull/15)

### 0.3.2 (2019-01-01)
* Revise test files' names. [#11](https://github.com/akito19/rack-joint/pull/11)
* Support for Ruby 2.6.0 and update gemspec. [#12](https://github.com/akito19/rack-joint/pull/12)

### 0.3.1 (2018-12-05)
* Enable to redirect with an only host replaced. [#9](https://github.com/akito19/rack-joint/pull/9)

### 0.3.0 (2018-09-01)
* Determine scheme in response to requests. [#7](https://github.com/akito19/rack-joint/pull/7)

### 0.2.0 (2018-08-30)
* Enable to use SSL when redirecting. [#4](https://github.com/akito19/rack-joint/pull/4)
* Refactor to run redirect process. [#5](https://github.com/akito19/rack-joint/pull/5)

### 0.1.1 (2018-08-27)
* Update a class name, method name and filename. [#1](https://github.com/akito19/rack-joint/pull/1)

### 0.1.0 (2018-08-25)
Initial :muscle:


test/test_helper.rb
$LOAD_PATH.unshift File.expand_path("../../lib", __FILE__)
require "rack/joint"
require "minitest/autorun"

def single_path_config
  Rack::Builder.parse_file('test/config/redirect_config.ru')
end

def multiple_paths_config
  Rack::Builder.parse_file('test/config/multiple_paths.ru')
end

def multiple_hosts_config
  Rack::Builder.parse_file('test/config/multiple_hosts.ru')
end

def only_host_config
  Rack::Builder.parse_file('test/config/change_host_with_each_path.ru')
end

def same_host_config
  Rack::Builder.parse_file('test/config/redirect_different_path_with_same_host.ru')
end

def same_url_config
  Rack::Builder.parse_file('test/config/set_same_url.ru')
end

def get_url_with_ssl_config
  Rack::Builder.parse_file('test/config/redirect_with_ssl.ru')
end

def redirect_all_paths
  Rack::Builder.parse_file('test/config/change_host_with_all_paths.ru')
end


test/rack/joint_test.rb
require_relative '../test_helper'
require 'rack/test'
require 'rack/joint'

class JointTest < MiniTest::Test
  include Rack::Test::Methods

  class SinglePathTest < JointTest
    def app
      single_path_config
    end

    def test_joint
      get '/foo/bar/baz', {}, 'HTTP_HOST' => 'example.com'
      assert_equal 301, last_response.status
      assert_equal 'http://example.org/path/to/blah', last_response['location']
      assert_equal 'Redirect from: http://example.com/foo/bar/baz', last_response.body
    end
  end

  class MultiplePathsTest < JointTest
    def app
      multiple_paths_config
    end

    def test_joint
      get '/dogs/bark.html', {}, 'HTTP_HOST' => 'example.com'
      assert_equal 301, last_response.status
      assert_equal 'http://example.org/bowwow/woof', last_response['location']
      assert_equal 'Redirect from: http://example.com/dogs/bark.html', last_response.body

      get '/cats/meow.html', {}, 'HTTP_HOST' => 'example.com'
      assert_equal 302, last_response.status
      assert_equal 'http://example.org/meow/meow/mew', last_response['location']
      assert_equal 'Redirect from: http://example.com/cats/meow.html', last_response.body

      get '/frog.html', {}, 'HTTP_HOST' => 'example.com'
      assert_equal 301, last_response.status
      assert_equal 'http://example.org/croak', last_response['location']
      assert_equal 'Redirect from: http://example.com/frog.html', last_response.body
    end
  end

  class MultipleHostsTest < JointTest
    def app
      multiple_hosts_config
    end

    def test_joint
      get '/dogs/bark.html', {}, 'HTTP_HOST' => 'example.net'
      assert_equal 301, last_response.status
      assert_equal 'http://example.org/bowwow/woof', last_response['location']
      assert_equal 'Redirect from: http://example.net/dogs/bark.html', last_response.body

      get '/cats/meow.html', {}, 'HTTP_HOST' => 'example.com'
      assert_equal 301, last_response.status
      assert_equal 'http://example.org/meow/meow/mew', last_response['location']
      assert_equal 'Redirect from: http://example.com/cats/meow.html', last_response.body
    end
  end

  class NotSetNewHostTest < JointTest
    def app
      same_host_config
    end

    def test_joint
      get '/dogs/bark.html', {}, 'HTTP_HOST' => 'example.com'
      assert_equal 301, last_response.status
      assert_equal 'http://example.com/bowwow/woof', last_response['location']
      assert_equal 'Redirect from: http://example.com/dogs/bark.html', last_response.body

      get '/cats/meow.html', {}, 'HTTP_HOST' => 'example.com'
      assert_equal 302, last_response.status
      assert_equal 'http://example.com/meow/meow/mew', last_response['location']
      assert_equal 'Redirect from: http://example.com/cats/meow.html', last_response.body
    end
  end

  class NotSetNewPathTest < JointTest
    def app
      only_host_config
    end

    def test_joint
      get '/foo', {}, 'HTTP_HOST' => 'example.com'
      assert_equal 301, last_response.status
      assert_equal 'http://example.org/foo', last_response['location']
      assert_equal 'Redirect from: http://example.com/foo', last_response.body
    end
  end

  class SetSameURLTest < JointTest
    def app
      same_url_config
    end

    def test_joint
      assert_raises Rack::Joint::BadRedirectError do
        get '/foo/bar/baz', {}, 'HTTP_HOST' => 'example.com'
      end
    end
  end

  class NoSetPathTest < JointTest
    def app
      single_path_config
    end

    def test_joint
      get '/path-has-not-set-in-config', {}, 'HTTP_HOST' => 'example.com'
      assert_equal 200, last_response.status
    end
  end

  class AccessWithSSLHostTest < JointTest
    def app
      get_url_with_ssl_config
    end

    def test_joint
      get '/foo/bar/baz', {}, 'HTTP_HOST' => 'example.com', 'HTTPS' => 'on'
      assert_equal 301, last_response.status
      assert_equal 'https://example.org/path/to/blah', last_response['location']
      assert_equal 'Redirect from: https://example.com/foo/bar/baz', last_response.body

      get '/dogs/bark.html', {}, 'HTTP_HOST' => 'example.com', 'HTTPS' => 'on'
      assert_equal 301, last_response.status
      assert_equal 'http://example.org/bowwow', last_response['location']
      assert_equal 'Redirect from: https://example.com/dogs/bark.html', last_response.body

      get '/cats/meow.html', {}, 'HTTP_HOST' => 'example.com', 'HTTPS' => 'on'
      assert_equal 301, last_response.status
      assert_equal 'https://example.org/meow', last_response['location']
      assert_equal 'Redirect from: https://example.com/cats/meow.html', last_response.body
    end
  end

  class RedirectAllPathsTest < JointTest
    def app
      redirect_all_paths
    end

    def test_joint
      get '/cats/meow.html', {}, 'HTTP_HOST' => 'example.com', 'HTTPS' => 'on'
      assert_equal 301, last_response.status
      assert_equal 'https://example.org/cats/meow.html', last_response['location']
      assert_equal 'Redirect from: https://example.com/cats/meow.html', last_response.body

      get '/dogs/bow', {}, 'HTTP_HOST' => 'example.com', 'HTTPS' => 'on'
      assert_equal 301, last_response.status
      assert_equal 'https://example.org/dogs/bow', last_response['location']
      assert_equal 'Redirect from: https://example.com/dogs/bow', last_response.body
    end
  end

  class RedirectWithQueryParameters < JointTest
    def app
      only_host_config
    end

    def test_joint
      get '/foo?hoge=abcdef&fuga=ghijkl', {}, 'HTTP_HOST' => 'example.com', 'HTTPS' => 'on'
      assert_equal 301, last_response.status
      assert_equal 'https://example.org/foo?hoge=abcdef&fuga=ghijkl', last_response['location']
      assert_equal 'Redirect from: https://example.com/foo?hoge=abcdef&fuga=ghijkl', last_response.body
    end
  end
end


lib/rack/joint.rb
require "rack"
require "rack/joint/context"
require "rack/joint/redirect"
require "rack/joint/redirect_interface"
require "rack/joint/url_builder"
require "rack/joint/version"

module Rack
  class Joint
    attr_reader :app, :block
    def initialize(app, &block)
      @app = app
      @block = block if block_given?
    end

    def call(env)
      rack_request = Rack::Request.new(env)
      mappings = Context.new(rack_request).instance_exec(&block)
      env_host = rack_request.host
      responses = valid_mapping(mappings, env_host)

      if check_redirect?(rack_request, responses)
        redirect_info(rack_request, responses)
      else
        app.call(env)
      end
    end

    private

    # @param mappings [Array] URI mappings for redirecting.
    # @param host    [String] Requested hostname(env).
    # @return [Array] Return URL mapped responses.
    def valid_mapping(mappings, host)
      return [] unless mappings
      mappings.select { |res| res[host] }.first.values.flatten(1)
    end

    # @param  request   [Rack::Request] Request env.
    # @param  responses [Array] Mapped redirect response information.
    # @return [Boolean]
    def check_redirect?(request, responses)
      current_url = request.url
      responses.map do |res|
        res[2].to_s.include?(current_url)
      end.include?(true)
    end

    # @param  request   [Rack::Request] Request env.
    # @param  responses [Array] Mapped redirect response information.
    # @return [Array] Return response corresponded request.
    def redirect_info(request, responses)
      current_url = request.url
      responses.select do |res|
        res[2].to_s.include?(current_url)
      end.to_a.flatten(1)
    end
  end
end


lib/rack/joint/url_builder.rb
# frozen_string_literal: true
require 'uri'

module Rack
  class Joint
    class UrlBuilder
      attr_reader :scheme, :query, :host, :path
      def initialize(scheme, request_query, host, path)
        @scheme = scheme
        @query = request_query
        @host = host
        @path = path
      end

      # @return [URI] URL with SSL or non-SSL
      def build
        if scheme == 'https'
          https_url_builder
        else
          http_url_builder
        end
      end

      private

      # @return [URI]
      def https_url_builder
        # When query parameters isn't added to request URL,
        # returns empty String object with `Rack::Request::Helper#query_string`.
        if query.empty?
          URI::HTTPS.build({ host: host, path: path }).to_s
        else
          URI::HTTPS.build({ host: host, path: path, query: query }).to_s
        end
      end

      # @return [URI]
      def http_url_builder
        # When query parameters isn't added to request URL,
        # returns empty String object with `Rack::Request::Helper#query_string`.
        if query.empty?
          URI::HTTP.build({ host: host, path: path }).to_s
        else
          URI::HTTP.build({ host: host, path: path, query: query }).to_s
        end
      end
    end
  end
end


lib/rack/joint/redirect_interface.rb
# frozen_string_literal: true
module Rack
  class Joint
    class BadRedirectError < StandardError; end

    class RedirectInterface
      attr_reader :scheme, :request_scheme, :request_query, :old_host, :old_path
      def initialize(request, old_host, old_path, &block)
        @status = 301
        @scheme = nil
        @request_query = request.query_string
        @request_scheme = request.scheme
        @old_host = old_host
        @old_path = old_path || request.path_info
        instance_exec(&block)
      end

      # @return [Array] Return response given parameters in `config.ru`.
      def apply!
        @scheme ||= request_scheme
        @new_host ||= old_host
        @new_path ||= old_path
        old_url = UrlBuilder.new(request_scheme, request_query, old_host, old_path).build
        new_location = UrlBuilder.new(scheme, request_query, @new_host, @new_path).build
        if old_url == new_location
          raise BadRedirectError.new('Redirect URL has been declared the same as current URL.')
        end
        [@status, { 'Location' => new_location, 'Content-Type' => 'text/html', 'Content-Language' => '0' }, ["Redirect from: #{old_url}"]]
      end

      private

      # @param flag [Boolean]
      # @return [String] `http` or `https`
      def ssl(flag)
        @scheme =
          if flag
            'https'
          else
            'http'
          end
      end

      # @param status [Integer] `status` parameter when redirecting in `config.ru`.
      # @return [Integer] Return status code.
      def status(status)
        @status =
          case status
          when 302
            302
          when 303
            303
          when 307
            307
          when 308
            308
          else
            301
          end
      end

      # @param host [String] `new_host` parameter to redirect hostname in `config.ru`.
      # @return [String] Return hostname where redirects to.
      def new_host(host)
        @new_host = host
      end

      # @param path [String] `to` parameter to redirect path in `config.ru`
      # @return [String] Return pathname where redirects to.
      def to(path)
        @new_path = path
      end
    end
  end
end

